### ISSUE TITLE
[question] What is the recommended way to get Conan to use an existing CMake toolchain file?

### ISSUE THREAD

**Issue by PolyPik** [conan-io/conan#8860]
- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

I recently wrote a CMake toolchain file for the Texas Instrument C2000 Code Generation Tools toolchain. After verifying that it works as intended, I sought to find a way to get this toolchain file to work with Conan.

The first thing I found in my web searches  that conan has tools for generating a CMake toolchain file on the fly and including it in the call to CMake. However, after going over the source code of the CMakeToolchain class it seems that I am unable to specify the path of the compiler executable. Unless there is some way for me to set the path of the compiler executable in the `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` variables, I cannot use the CMakeToolchain toolchain generator.

Further research led me to the `CONAN_CMAKE_TOOLCHAIN_FILE` variable and CMake build helper. Defining the `CONAN_CMAKE_TOOLCHAIN_FILE` supposedly inserts a `-DCMAKE_TOOLCHAIN_FILE` option into the CMake command line. However, this can be rather inflexible if one is working with multiple toolchains. In the CMake build helper class, there is the `configure` method which has the defs parameter where I can pass a dictionary with a key `CMAKE_TOOLCHAIN_FILE` and the corresponding value being that of the path to the toolchain file. This looks to be the most promising as I can pick out which toolchain to use based on the Conan settings.

If there are any other ways to use a handwritten CMake toolchain file in Conan, I would like to have them described in a reply to this post.

**Comment by memsharded** [comment]
Hi @PolyPik 

The ``CMakeToolchain`` just got a major refactor and modularization, ready to be released for next 1.36 next week. Check https://github.com/conan-io/conan/pull/8749 for details about the implementation.
For example for the Android, it will use the AndroidNDK toolchain explicitly included as:
```cmake
        # New toolchain things
        set(ANDROID_PLATFORM {{ CMAKE_SYSTEM_VERSION }})
        set(ANDROID_STL {{ CMAKE_ANDROID_STL_TYPE }})
        set(ANDROID_ABI {{ CMAKE_ANDROID_ARCH_ABI }})
        include({{ CMAKE_ANDROID_NDK }}/build/cmake/android.toolchain.cmake)
```

We can easily add, and it is an already existing feature request, user toolchains. I would like to know a bit more about what UX would you prefer/envision:
- Explicitly adding the toolchain in the package files, and having something like ``toolchain.include = ["mytoolchain.cmake"]``
- It is possible, and very easy to directly define the templates, the cmake code directly to add new things there.
- Adding them at the [conf] level, because they will be global in the system or something (it is recommended to included them inside the packages for reproducibility), you can see for example how the AndroidNDK location is defined in current code:
```python
        android_ndk = self._conanfile.conf["tools.android"].ndk_path
        android_ndk = android_ndk.replace("\\", "/")
        ctxt_toolchain = {
            'CMAKE_ANDROID_NDK': android_ndk,
        }
```

Some hints:
- You should be looking at new ``tools.cmake.xxxx`` things, they are still experimental, but they are the future of Conan 2.0 and major modifications and effort is being done here.
- Env-vars like CONAN_CMAKE_TOOLCHAIN_FILE  will not be used here (they are global state, which is difficult to control).
- Generation and customization of Conan toolchains, like CMakeToolchain is to be done in ``generate()`` method. It should be easy to extend these with completely new "blocks" or to modify existing blocks, both python logic and the cmake templates.

Please let me know.

**Comment by PolyPik** [comment]
Hi memshard,

Thanks for the quick feedback. For UX preference, in order of descending preference I would order it Option 3, Option 1, and Option 2.

I am of the opinion that toolchain files are rarely ever changed and any changes to a toolchain file should affect all projects that rely on that toolchain file. Also, I am not that fond of using templates unless I absolutely have to (e.g. websites).

Anyhow, I am happy to hear that improvements to the CMakeToolchain will be released in the near future.

**Comment by memsharded** [comment]
Hi @PolyPik 

Proposing this PR https://github.com/conan-io/conan/pull/8945 for next 1.37, implementing option 3, and also the possibility of completely overriding the CMakeToolchain ``conan_toolchain.cmake`` file. Feedback welcome!

**Comment by jsanchez19** [comment]
@PolyPik  - I ran across this issue while searching for an example on packaging a toolchain file onto a Conan package for re-usability. I'm gathering that this is perhaps not supported in current conan. Coincidentally, I'm looking to do this for the TI c2000 toolchain too. I already have a working `toolchain.cmake` file as well and I was trying to follow the "[cross building](https://docs.conan.io/en/latest/systems_cross_building/cross_building.html#using-build-requires)" documentation but haven't had much success yet. 

If you have a working setup I'd love to learn how you got it working.
 
My goal was to be able to add a build_requires to a microcontroller project that would import the toolchain package and build with it. Something like the profile file below (from [here](https://docs.conan.io/en/latest/integrations/cross_platform/ios.html#darwin-toolchain)):

**ti_c2000_profile:**
```
include(default)

[settings]
os=none
arch=c2000

[build_requires]
ti-c2000-toolchain/1.0
```

**Comment by memsharded** [comment]
I want to give a heads up, besides https://github.com/conan-io/conan/pull/8945 we are also exploring how build_requires and other packages could transmit configuration information to its dependents. 

There is already a proof of concept in https://github.com/conan-io/conan/pull/8895, please track it to see progress.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/8945 merged, adding custom user toolchains will be released in 1.37
